components/de_amp/browser/de_amp_throttle.cc

```diff
@@ -0,0 +1,116 @@
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/page_navigator.h"
+#include "content/public/browser/web_contents.h"
+#include "mojo/public/cpp/bindings/pending_receiver.h"
+#include "mojo/public/cpp/bindings/pending_remote.h"
+#include "net/base/net_errors.h"
+#include "services/network/public/mojom/url_loader.mojom.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+#include "ui/base/page_transition_types.h"
+#include "ui/base/window_open_disposition.h"
+
+namespace de_amp {
+
+// static
+std::unique_ptr<DeAmpThrottle> DeAmpThrottle::MaybeCreateThrottleFor(
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
+    const network::ResourceRequest& request,
+    const content::WebContents::Getter& wc_getter) {
+  auto* contents = wc_getter.Run();
+
+  if (!contents)
+    return nullptr;
+
+  PrefService* prefs =
+      user_prefs::UserPrefs::Get(contents->GetBrowserContext());
+
+  if (!base::FeatureList::IsEnabled(de_amp::features::kBraveDeAMP) ||
+      !prefs->GetBoolean(kDeAmpPrefEnabled)) {
+    return nullptr;
+  }
+
+  return std::make_unique<DeAmpThrottle>(task_runner, request, wc_getter);
+}
+
+DeAmpThrottle::DeAmpThrottle(
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
+    const network::ResourceRequest& request,
+    const content::WebContents::Getter& wc_getter)
+    : task_runner_(task_runner), request_(request), wc_getter_(wc_getter) {}
+
+DeAmpThrottle::~DeAmpThrottle() = default;
+
+void DeAmpThrottle::WillProcessResponse(
+    const GURL& response_url,
+    network::mojom::URLResponseHead* response_head,
+    bool* defer) {
+  VLOG(2) << "deamp throttling: " << response_url;
+  *defer = true;
+
+  mojo::PendingRemote<network::mojom::URLLoader> new_remote;
+  mojo::PendingReceiver<network::mojom::URLLoaderClient> new_receiver;
+  mojo::PendingRemote<network::mojom::URLLoader> source_loader;
+  mojo::PendingReceiver<network::mojom::URLLoaderClient> source_client_receiver;
+  DeAmpURLLoader* de_amp_loader;
+  std::tie(new_remote, new_receiver, de_amp_loader) =
+      DeAmpURLLoader::CreateLoader(weak_factory_.GetWeakPtr(), response_url,
+                                   task_runner_);
+  InterceptAndStartLoader(
+      std::move(source_loader), std::move(source_client_receiver),
+      std::move(new_remote), std::move(new_receiver), de_amp_loader);
+}
+
+void DeAmpThrottle::Redirect(const GURL& new_url) {
+  delegate_->CancelWithError(net::ERR_ABORTED);
+
+  auto* contents = wc_getter_.Run();
+
+  if (!contents)
+    return;
+
+  auto* entry = contents->GetController().GetVisibleEntry();
+  if (!entry)
+    return;
+
+  content::OpenURLParams params(
+      new_url,
+      content::Referrer::SanitizeForRequest(new_url, entry->GetReferrer()),
+      contents->GetMainFrame()->GetFrameTreeNodeId(),
```

Comment:
yes, this is the equivalent of calling `FromNavigationHandle`, but we don't have access to the navigation handle in a URLLoaderThrottle

Reply:
The tab can change the content (and mainFrame) after this point and before we call `web_contents->OpenURL()` (for example because of bf cache) => saved here `FrameTreeNodeId` != real `contents->GetMainFrame()->GetFrameTreeNodeId()`.
What happens in this case?

Reply:
@atuchin-m it doesn't seem to be a problem in the dozen or so places that use `FromNavigationHandle` to navigate to a new url

Reply:
@bridiver if we called `web_contents->OpenURL()` here it will be okay. But we call `OpenURL()` via **PostTask** => web_contents and frame_tree can be changed.
We we need to explicitly specify `frame_tree_node_id`?

Reply:
@atuchin-m the post task was copied from `TypedNavigationUpgradeThrottle::OnHttpsLoadTimeout` 

```// Post a task to navigate to the fallback URL. We don't navigate
  // synchronously here, as starting a navigation within a navigation is
  // an antipattern.```

Reply:
Discussed in DM. For main frames it should always be `RenderFrameHost::kNoFrameTreeNodeId`, so doesn't matter.